src/libraries/System.Private.CoreLib/src/System/Globalization/GlobalizationMode.Unix.cs

```diff
@@ -7,42 +7,48 @@ internal static partial class GlobalizationMode
     {
         private static partial class Settings
         {
-            internal static readonly bool Invariant = GetGlobalizationInvariantMode();
-        }
-
-        internal static bool Invariant => Settings.Invariant;
-
-        internal static bool UseNls => false;
+            /// <summary>
+            /// Load ICU (when not in Invariant mode) in a static cctor to ensure it is loaded early in the process.
+            /// Other places, e.g. CompareInfo.GetSortKey, rely on ICU already being loaded before they are called.
+            /// </summary>
+            static Settings()
+            {
+                if (!Invariant)
+                {
+                    if (TryGetAppLocalIcuSwitchValue(out string? icuSuffixAndVersion))
+                    {
+                        LoadAppLocalIcu(icuSuffixAndVersion);
+                    }
+                    else
+                    {
+                        int loaded = LoadICU();
+                        if (loaded == 0)
+                        {
+                            Environment.FailFast(GetIcuLoadFailureMessage());
+                        }
+                    }
+                }
+            }
 
-        private static bool GetGlobalizationInvariantMode()
-        {
-            bool invariantEnabled = GetInvariantSwitchValue();
-            if (!invariantEnabled)
+            private static string GetIcuLoadFailureMessage()
             {
-                if (TryGetAppLocalIcuSwitchValue(out string? icuSuffixAndVersion))
+                // These strings can't go into resources, because a resource lookup requires globalization, which requires ICU
+                if (OperatingSystem.IsBrowser() || OperatingSystem.IsIOS() || OperatingSystem.IsAndroid())
```

Comment:
I'd go with the simple message everywhere. Otherwise, you would also need to check for tvOS, MacCatalyst, macOS.

Reply:
I'll add the check for `tvOS` and `MacCatalyst`. My concern with using the "simple" message is that for "desktop" scenarios like Linux and macOS, it would be less informative than it is today.

Reply:
I don't think macOS users will know what to do with the message and in the case of macOS UI apps it makes even less sense if the user somehow manages to see it.

Reply:
I believe it does make sense for macOS users. We require they have ICU on the machine. We don't bring ICU with .NET on macOS.

See https://docs.microsoft.com/en-us/dotnet/standard/globalization-localization/globalization-icu#macos-behavior.

Reply:
@marek-safar - I'm going to merge this for now. If we decide to do something different here for macOS, we can open a new issue/PR for it. But, as-is, this change is keeping the current/existing behavior for macOS.